File: golang/139.word-break.go
/*
 * @lc app=leetcode id=139 lang=golang
 *
 * [139] Word Break
 */
package main

import "fmt"

// @lc code=start
// to be honest, it is not a typic dynamic programming
// a important key to solve this problem is contraints 3 that wordDict[i].lenght <=20
func wordBreak(s string, wordDict []string) bool {
	dp := []bool{true}
	for i := 0; i < len(s); i++ {
		matched := false
		for j := 0; j < len(wordDict); j++ {
			word := wordDict[j]
			length := len(wordDict[j])
			if i-length+1 >= 0 && i-length < len(dp) && dp[i-length+1] {
				if s[i-length+1:i+1] == word {
					matched = true
					break
				}
			}
		}
		dp = append(dp, matched)
	}
	fmt.Println(dp)
	return dp[len(s)]
}

// @lc code=end

func main() {
	fmt.Println(wordBreak("leetcode", []string{"leet", "code"}))
}

File: cpp/139.word-break.cpp
/*
 * @lc app=leetcode id=139 lang=cpp
 *
 * [139] Word Break
 */
#include<leetcode-helper.h>
// @lc code=start
class Solution {
public:
    bool wordBreak(string s, vector<string>& wordDict) {
        // default??
        bool dp[s.size()];
        dp[0] = true;
        for (int i = 0;i<s.size();i++){
            for (auto word:wordDict){
                // TODO
                if (i-word.size()>0&&dp[i-word.size()]&&s.substr(i-word.size(),i)==word){
                    dp[i] = true;
                    continue;
                }
            }
        }
        return dp[s.size() - 1];
    }
};
// @lc code=end

